utilities/basic-utils/src/main/kotlin/NativeMemoryAllocator.kt

```diff
@@ -6,15 +6,37 @@
 package org.jetbrains.kotlin.konan.util
 
 import sun.misc.Unsafe
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicStampedReference
+
+@PublishedApi
+internal val allocatorHolder = AtomicStampedReference<NativeMemoryAllocator>(null, 0)
+
+@PublishedApi
+internal var counter = AtomicInteger(0)
+
+inline fun <R> usingNativeMemoryAllocator(block: () -> R): R {
+    counter.getAndIncrement()
+    do {
+        val reference = allocatorHolder.reference
+        val stamp = allocatorHolder.stamp
+        val allocator = reference ?: NativeMemoryAllocator()
+    } while (!allocatorHolder.compareAndSet(reference, allocator, stamp, stamp + 1))
```

Comment:
Using non-atomic counter and holder under single lock should be much simpler and more reliable.
Also similar data race probably needs to be fixed in JvmCallbacks.